Add delayed evaluation functionality

brkronheim opened the pull request:
This pull request adds the ability to delay the calculation of branches based on the nominal tree until after an event has been selected to go into a histogram.

The main use for this are situations where you need to calculate some relatively expensive quantity based on GEN/LHE level variables such as a weight, and would like to do so a minimum number of times. Currently, the only ways to do this are in process_extra_after_presel or process_extra_after_skim. Doing so after preselection has the disadvantage of running over all passing events for each shape variation, while doing so after the skim results in many events which will be cut later being processed. 

This adds an object to the base processor class which will be made in the base processor constructor. In the constructor for a user processor class, the user can add lines such as 

```
self.delayed_branches.register("drWeight", dr_fn, default_value=1.0)
self.delayed_branches.register("stitchWeight", stitch_fn, default_value=1.0)
self.delayed_branches.register(
    [
        "nloBDTScoreMean",
        "nloBDTScoreUncert",
        "nloReweight",
        "nloReweightUp",
        "nloReweightDown",
    ],
    nlo_group_fn,
    default_value=1.0,
)
```

where a branch name (or list of branch names) is specified, a function is provided, and a default value is provided. 

When the nominal variation is run, an index for all the events is created, and a branch for each created variable is stored locally with value equal to the default specified when the branch is registered. Further, a boolean array is initialized to False for all events, indicating that none of the events have passed any selections yet.

Then, after the categories have been defined, an update for the branches is called. This identifies all events which pass at least one of the category selections. It then compares this list with boolean array stored in the object to identify newly passing events. These are then fed into the stored functions for each branch and the output is used to update the local branches for each object. These are then used to update the main events Awkward Array.

The result is that the functions are run exactly once for only those events which pass at least one category selection, and they are ready by the time compute_weights is called.


I'm unsure if you need me to run or add tests for this, do a specific documentation update, or change where/how this is integrated into the framework, but I'll be happy to make any adjustments you need before integrating this.

brkronheim committed:
add delayed eval

brkronheim commented:
Hi @valsdav , thanks for looking at this! I would be happy to add a test config, though I won't be able to until next Monday. I'll add one then, and if you have any other questions/comments I can address them then as well

brkronheim committed:
Merge branch 'PocketCoffea:main' into delayedEval

brkronheim committed:
Add delayed eval test

brkronheim commented:
Hi @valsdav, I've added the test you requested. 

The test defines a parity branch that is only computed for events ending up in histograms. The branch is initialized to -1 for all events. It is then set to 0 for event events and 1 to odd events. The test defines a region based off of the parity, calculated at a different stage in the workflow. It then histograms the delayed parity branch for each region and confirms that only the appropriate histogram bins are non-zero.

It also defines a systematic weight based on the parity branch. It then confirms that the branch has been correctly defined at this stage.

In total, this test checks that the delayed branch is always calculated for passing events and that it is calculated in time to define weights and fill histograms.

brkronheim commented:
Hi @valsdav, do you need any additional tests/documentation added to this?

valsdav commented on pocket_coffea/lib/delayed_eval.py:
```diff
@@ -0,0 +1,128 @@
+            not_done_mask = ~self.computed_nom_mask[nom_idxs_pass]
+            nom_idxs_new = np.unique(nom_idxs_pass[not_done_mask])
+            if nom_idxs_new.size > 0:
+                ev_sub = self.nominal_events[nom_idxs_new]
```
HI! Can you please guide me a bit about what's happening in this section? I'm trying to understand why the processor object is needed on the delayedeval class side

brkronheim committed:
Merge branch 'main' into delayedEval

brkronheim committed:
update to allow multi dim variables

brkronheim commented:
Hi @valsdav, I'd like to try and get this integrated, as it would benefit the runtime of our analysis a lot to be able to use this. I removed the unnecessary passing of the procesor object and also updated the caching to handle multi dimensional arrays, and updated the testing a bit. All the tests pass as the moment:

```
[bkronhei@lxplus963 PocketCoffea]$ .venv/bin/pytest tests/test_full_configs/test_full_configs.py 
========================================== test session starts ===========================================
platform linux -- Python 3.11.13, pytest-9.0.3, pluggy-1.6.0
rootdir: /afs/cern.ch/user/b/bkronhei/PoCoAnalysis/PocketCoffea
configfile: pyproject.toml
plugins: typeguard-4.5.1, xdist-3.8.0, cov-7.1.0
collected 12 items                                                                                       

tests/test_full_configs/test_full_configs.py ............                                          [100%]

============================================ warnings summary ============================================
.venv/lib64/python3.11/site-packages/awkward/_typeparser/generated_parser.py:318
  /afs/cern.ch/user/b/bkronhei/PoCoAnalysis/PocketCoffea/.venv/lib64/python3.11/site-packages/awkward/_typeparser/generated_parser.py:318: DeprecationWarning: module 'sre_parse' is deprecated
    import sre_parse

.venv/lib64/python3.11/site-packages/awkward/_typeparser/generated_parser.py:319
  /afs/cern.ch/user/b/bkronhei/PoCoAnalysis/PocketCoffea/.venv/lib64/python3.11/site-packages/awkward/_typeparser/generated_parser.py:319: DeprecationWarning: module 'sre_constants' is deprecated
    import sre_constants

.venv/lib64/python3.11/site-packages/uproot/extras.py:15
  /afs/cern.ch/user/b/bkronhei/PoCoAnalysis/PocketCoffea/.venv/lib64/python3.11/site-packages/uproot/extras.py:15: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
    import pkg_resources

tests/test_full_configs/test_full_configs.py: 10 warnings
  /afs/cern.ch/user/b/bkronhei/PoCoAnalysis/PocketCoffea/.venv/lib64/python3.11/site-packages/coffea/nanoevents/schemas/nanoaod.py:205: RuntimeWarning: Missing cross-reference index for LowPtElectron_electronIdx => Electron
    warnings.warn(

tests/test_full_configs/test_full_configs.py: 10 warnings
  /afs/cern.ch/user/b/bkronhei/PoCoAnalysis/PocketCoffea/.venv/lib64/python3.11/site-packages/coffea/nanoevents/schemas/nanoaod.py:205: RuntimeWarning: Missing cross-reference index for LowPtElectron_photonIdx => Photon
    warnings.warn(

tests/test_full_configs/test_full_configs.py: 21 warnings
  /afs/cern.ch/user/b/bkronhei/PoCoAnalysis/PocketCoffea/pocket_coffea/lib/jets.py:363: PydanticDeprecatedSince20: The `parse_raw` method is deprecated; if your data is JSON use `model_validate_json`, otherwise load the data then use `model_validate` instead. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.13/migration/
    cset = CorrectionSet.parse_raw(fin.read())

tests/test_full_configs/test_full_configs.py: 21 warnings
  /afs/cern.ch/user/b/bkronhei/PoCoAnalysis/PocketCoffea/.venv/lib64/python3.11/site-packages/pydantic/main.py:1415: PydanticDeprecatedSince20: `load_str_bytes` is deprecated. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.13/migration/
    obj = parse.load_str_bytes(

tests/test_full_configs/test_full_configs.py: 21 warnings
  /afs/cern.ch/user/b/bkronhei/PoCoAnalysis/PocketCoffea/pocket_coffea/lib/jets.py:376: PydanticDeprecatedSince20: The `parse_obj` method is deprecated; use `model_validate` instead. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.13/migration/
    res = Correction.parse_obj(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
============================== 12 passed, 86 warnings in 650.70s (0:10:50) ===============================
```

valsdav commented:
hi @brkronheim ! Thanks for the changes. I haven't looked at it in details but I see that it's quite orthogonal to the nominal processing. It's good to go for me!
Thanks

brkronheim commented:
Thanks @valsdav! Do you have an idea on how long it will take for this update to make its way into the latest tag of the container on cvmfs?

valsdav commented:
The pipeline for building the image is ongoing: https://gitlab.cern.ch/cms-analysis/general/PocketCoffea/-/pipelines
Then the unpacker will put it on cvmfs. I think few hours
